### PR TITLE
SC-718 BigInt

### PR DESCRIPTION
**Pull request by potan**

### CODE REVIEW

**Review by msmolyakov — APPROVED** [review]

File: lang/tests/src/test/scala/com/wavesplatform/lang/IntegrationTest.scala [hunk 1]
```diff
@@ -1364,6 +1364,18 @@ class IntegrationTest extends PropSpec with PropertyChecks with ScriptGen with M
     eval(src, version = V4) should produce("Compilation failed: [Non-matching types: expected: Int, actual: String")
   }
 
+  property("List[Int] median - list with big elements - error") {
+    val src =
+      s"""[${Long.MaxValue}, ${Long.MaxValue-2}].median()"""
+    eval(src, version = V4) shouldBe Right(CONST_LONG(Long.MaxValue-1))
```
Comment (xrtm000) [inline lang/tests/src/test/scala/com/wavesplatform/lang/IntegrationTest.scala:1390]: Fix of median should be started from V5

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala [hunk 2]
```diff
@@ -95,6 +99,96 @@ object PureContext {
...
+  lazy val bigIntToString: BaseFunction[NoContext] =
+    NativeFunction("toStringBigInt", 65, BIGINT_TO_STRING, STRING, ("n", BIGINT)) {
+      case CONST_BIGINT(n) :: Nil => CONST_STRING(n.toString)
+      case xs => notImplemented[Id, EVALUATED]("toString(n: BigInt)", xs)
+    }
+
+  lazy val stringToBigInt: BaseFunction[NoContext] =
+    NativeFunction("parseBigIntValue", 65, STRING_TO_BIGINT, BIGINT, ("n", STRING)) {
+      case CONST_STRING(n) :: Nil => Either.cond(n.length <= 155 ,BigInt(n), s"String too long for 512-bits big integers (${n.length} when max is 155)")
+        .filterOrElse(v => v <= maxBigInt && v >= minBigInt, "Value to big for 512-bits big integer")
```
Comment (xrtm000) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala]:
```suggestion
        .filterOrElse(v => v <= maxBigInt && v >= minBigInt, "Value too big for 512-bits big integer")
```

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala [hunk 3]
```diff
@@ -95,6 +99,96 @@ object PureContext {
...
+        .map(CONST_BIGINT.apply)
+      case xs => notImplemented[Id, EVALUATED]("parseBigIntValue(n: String)", xs)
+    }
+
+  lazy val stringToBigIntOpt: BaseFunction[NoContext] =
+    NativeFunction("parseBigInt", 65, STRING_TO_BIGINTOPT, BIGINT, ("n", STRING)) {
```
Comment (xrtm000) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala]:
```suggestion
    NativeFunction("parseBigInt", 65, STRING_TO_BIGINTOPT, UNION(BIGINT, UNIT), ("n", STRING)) {
```

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala [hunk 4]
```diff
@@ -95,6 +99,96 @@ object PureContext {
...
+
+  lazy val bigIntToBytes: BaseFunction[NoContext] =
+    NativeFunction("toBytesBigInt", 65, BIGINT_TO_BYTES, BYTESTR, ("n", BIGINT)) {
+      case CONST_BIGINT(n) :: Nil => CONST_BYTESTR(ByteStr(n.toByteArray))
+      case xs => notImplemented[Id, EVALUATED]("toBytesBigInt(n: BigInt)", xs)
+    }
+
+  lazy val bytesToBigIntLim: BaseFunction[NoContext] =
+    NativeFunction("toBigInt", 65, BYTES_TO_BIGINT_LIM, BIGINT, ("n", BYTESTR), ("off", LONG), ("size", LONG)) {
+      case CONST_BYTESTR(ByteStr(n)) :: CONST_LONG(off) :: CONST_LONG(s) :: Nil => Either.cond(off >= 0 && off <= n.size && s <= 64 && s > 0, CONST_BIGINT(BigInt(n.slice(off.toInt, (off+s).toInt))), s"ByteStr too long (s{s} > 64 bytes)")
```
Comment (xrtm000) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala]:
```suggestion
      case CONST_BYTESTR(ByteStr(n)) :: CONST_LONG(off) :: CONST_LONG(s) :: Nil => Either.cond(off >= 0 && off <= n.size && s <= 64 && s > 0, CONST_BIGINT(BigInt(n.slice(off.toInt, (off+s).toInt))), s"ByteStr too long ($s > 64 bytes)")
```

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala [hunk 5]
```diff
@@ -95,6 +99,96 @@ object PureContext {
...
+      case xs => notImplemented[Id, EVALUATED]("toBigInt(n: ByteStr, offset: Int, size: Int)", xs)
+    }
+  lazy val bytesToBigInt: BaseFunction[NoContext] =
+    NativeFunction("toBigInt", 65, BYTES_TO_BIGINT, BIGINT, ("n", BYTESTR)) {
+      case CONST_BYTESTR(ByteStr(n)) :: Nil => Either.cond(n.size <= 64, CONST_BIGINT(BigInt(n)), s"ByteStr too long (s{n.size} > 64 bytes)")
```
Comment (xrtm000) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/PureContext.scala]:
```suggestion
      case CONST_BYTESTR(ByteStr(n)) :: Nil => Either.cond(n.size <= 64, CONST_BIGINT(BigInt(n)), s"ByteStr too long (${n.size} > 64 bytes)")
```

**Review by xrtm000 — CHANGES_REQUESTED** [review]
Add tests for `toBytesBigInt`, check that `powBigInt` and `logBigInt` don't allow to exceed BigInt limits

File: lang/doc/v5/funcs/operators.hjson [hunk 6]
```diff
@@ -1,5 +1,19 @@
...
+      name: "+"
+      params: [ "Int", "Int" ]
+      doc: "Sum integers."
+      paramsDoc: [ "First value", "Second value" ]
+      complexity: 1
+    }
+    {
+      name: "-"
+      params: [ "Int", "Int" ]
+      doc: "Substitute integers."
```
Comment (alexeykiselev) [inline lang/doc/v5/funcs/operators.hjson]: `Substitute integers` -> `Subtract integers`

File: lang/doc/v5/funcs/operators.hjson [hunk 7]
```diff
@@ -23,17 +37,38 @@
...
       paramsDoc: [ "First value", "Second value" ]
-      complexity: 1
+      complexity: 8
     }
     {
       name: "-"
-      params: [ "Int", "Int" ]
-      doc: "Substitute integers."
+      params: [ "BigInt", "BigInt" ]
+      doc: "Substitute big integers."
```
Comment (alexeykiselev) [inline lang/doc/v5/funcs/operators.hjson]: Substitue -> Subtract

**Review by alexeykiselev — CHANGES_REQUESTED** [review]

**Review by xrtm000 — APPROVED** [review]

**Review by mrkraft — APPROVED** [review]

**Pull request merged**